csrc/device_lower/analysis/circular_buffer.cpp

```diff
@@ -286,6 +296,44 @@ void CircularBufferInfo::setCircularBufferOptions(
   }
 }
 
+// Derive number of computation warp groups from loop domain of its consumer
+// For example, in normalization kernel:
+// circular buffer: Ts[BIDy, Circular(S), compute groups(S),...], ca 2
+// its consumer is: Tl[BIDy, Circular(S), compute groups(TIDy),...], ca 3
+// Look for the corresponding loop domain in its consumer and make sure it
+// is parallelized same as warp specialized dim. Then, its content represents
+// number of computation warp groups.
+
+// TODO: Another approach is we can just save `computation_warp_groups_`
+// in  `circularBufferOptions` since it is a scheduler parameter.
+void CircularBufferInfo::setComputationWarpGroups(const TensorView* tv) {
+  auto option = tv->circularBufferOptions();
+  auto old_val = computation_warp_groups_;
+  // -1 indicates not set yet, if set, should not change
+  int64_t new_val = (old_val == -1) ? 1 : old_val;
+  if (std::holds_alternative<WarpSpecialized>(option.type)) {
+    auto ws_pt = std::get<WarpSpecialized>(option.type).on;
+    int64_t next_pos = getInnerMostCircularBufferPosition(tv) + 1;
+    auto consumer = ir_utils::consumerTvsOf(tv).at(0);
+    if (consumer->nDims() > next_pos &&
```

Comment:
Isn't the Idmodel more robust at checking if two IterDomains are equivalent?

Reply:
revised with IdModel